_posts/2021-08-15-interviews.markdown

```diff
@@ -0,0 +1,155 @@
+
+## Enter the LinkedList Pipeline
+
+Let's rewrite how we use this `LinkedList` class into something a little more _modern_:
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=pipeline1.yaml" type="text/javascript"></script>
+
+YAML! Nice. Yaml is fun to joke about, but Vicki Boykis posted [a twitter poll](https://twitter.com/vboykis/status/1407821631997759488) about which languages people use in Data Science/ML and the most common reply was "does yaml count?" If everyone asks, it means yes.
+
+The above yaml does not actually do anything, and this is ostensibly a post about software coding interviews, so let's build something that reverses a linked list given this pipeline definition. Is this overkill for the goal of reversing a 4-element linked list? Sure. But it's a good analog to the modular systems that we are paid to build and use as software engineers (especially Data/ML engineers). It's the kind of thing that we actually do every day, and should be familiar with. In other words, **Components are the new Pointers.**
+
+So let's dive in.
+
+### Coding problem: Implement a framework to execute this pipeline
+
+First we'll need to define a `Task` class to carry out the logic of each `task` defined in the yaml. A `Task` may take an input `LinkedList` from a previous `Task`, performs some operation on it, and produces an output. Here's the interface:
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=base_task.py" type="text/javascript"></script>
+
+It takes an input `LinkedList` and a `name` for the task, and has an abstract `execute` method that will have to mutate `self._lst` (which is maybe not ideal but we'll work with it). Let's extend `Task` to perform the actions seen in the yaml: `push_values`, `reverse_list`, and `print_list`.
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=tasks.py" type="text/javascript"></script>
+
+You can see that I moved the `reverse` method into `ReverseListTask`, so we can remove the method from the `LinkedList` class. One benefit of this task-based approach is that the logic that we want to perform is located within the `Task`, and this helps us achieve that task. None of the other `Task`s need to know about reversing a list, and having code around to do unnecessary tasks might become step towards "Dependency Hell."
+
+Our yaml config has an `action` field that we can use to decide which of these subclasses to use for each step in the pipeline. Let's use a factory function for that <sup id="a1">[1](#f1)</sup>.
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=task_factory.py" type="text/javascript"></script>
```

Comment:
I was going to post a comment here making a suggestion on the interface you have in your Task class, which was: would it be easier if instead of each Task having a reference to `lst`, you reworked the interface so the `execute` method received the `lst` to operate on? Then you don't have to pass `lst` to the TaskFactory so it can pass it to each task ...

but this made me think of a broader point you could perhaps work into the post - that by giving the candidate a more concrete task where they have much more of an opportunity to express themselves and be creative, it brings up the chance to discuss interesting aspects of the code they wrote that you don't get when they are just mechanically implementing a algorithm - like, why did you design the interface so each `Task` has a reference to `lst` in the constructor, versus this other idea? what are the trade-offs of each? etc. 

in other words, it'd be interesting to hammer a bunch in the post on why this approach is better - its more faithful to the actual work the candidate will do, it creates plenty of chances for discussion points, etc